src/main/java/org/opensearch/flowframework/transport/handler/SearchHandler.java

```diff
@@ -83,16 +102,40 @@ public void validateRole(
             // Case 2: If Security is enabled and filter is disabled, proceed with search as
             // user is already authenticated to hit this API.
             // case 3: user is admin which means we don't have to check backend role filtering
-            client.search(request, ActionListener.runBefore(listener, context::restore));
+            doSearch(request, tenantId, ActionListener.runBefore(listener, context::restore));
         } else {
             // Security is enabled, filter is enabled and user isn't admin
             try {
                 ParseUtils.addUserBackendRolesFilter(user, request.source());
                 logger.debug("Filtering result by {}", user.getBackendRoles());
-                client.search(request, ActionListener.runBefore(listener, context::restore));
+                doSearch(request, tenantId, ActionListener.runBefore(listener, context::restore));
             } catch (Exception e) {
                 listener.onFailure(e);
             }
         }
     }
+
+    private void doSearch(SearchRequest request, String tenantId, ActionListener<SearchResponse> listener) {
+        SearchDataObjectRequest searchRequest = SearchDataObjectRequest.builder()
+            .indices(request.indices())
+            .tenantId(tenantId)
+            .searchSourceBuilder(request.source())
+            .build();
+        sdkClient.searchDataObjectAsync(searchRequest).whenComplete((r, throwable) -> {
+            if (throwable == null) {
+                try {
+                    SearchResponse searchResponse = SearchResponse.fromXContent(r.parser());
+                    logger.info(Arrays.toString(request.indices()) + " search complete: {}", searchResponse.getHits().getTotalHits());
+                    listener.onResponse(searchResponse);
+                } catch (Exception e) {
+                    logger.error("Failed to parse search response", e);
+                    listener.onFailure(new FlowFrameworkException("Failed to parse search response", INTERNAL_SERVER_ERROR));
+                }
+            } else {
+                Exception cause = SdkClientUtils.unwrapAndConvertToException(throwable);
+                logger.error(Arrays.toString(request.indices()) + " search failed", cause);
```

Comment:
nit:
logger.error("Search failed for indices: {}", Arrays.toString(request.indices()), cause); instead of concatenation